### PR TITLE
Change WASM direct heap access to use helper functions

### PR DESCRIPTION
**Pull request by kg**
According to the ECMAScript spec, [the left-hand side of an assignment is evaluated before the right-hand side](https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation). This means that for the following statement:

```Module.HEAPU32[offset] = get_value();```

The write will be discarded if the heap grows during the call to ```get_value```. If the offset itself is a property with an accessor or the return value of a function call the problem applies there too. This is because heap growth requires creating a new set of typed arrays pointing at the new larger heap buffer, and the old arrays become 'detached', and all reads/writes to a detached buffer will silently fail.

Since almost any JS expression can secretly be invoking a property getter, the best solution for this is to do roughly what blazor does, and perform memory accesses through functions so that we can be certain the offset and value will be evaluated *before* the target is evaluated. With this approach, the above becomes:

```INTERNAL.setU32(offset, get_value());```

### CODE REVIEW

**Comment by kg** [comment]
I spent a bit trying to use BigInt to properly implement I64/U64 because Emscripten's setvalue/getvalue for them turn out to be broken, but it seems to not be possible to do that without breaking existing code.

File: src/mono/wasm/runtime/exports.ts [hunk 1]
```diff
@@ -252,6 +253,8 @@ export const INTERNAL: any = {
     mono_wasm_raise_debug_event,
     mono_wasm_runtime_is_ready: runtimeHelpers.mono_wasm_runtime_is_ready,
 };
+// Expose all the get/set methods via INTERNAL
+Object.assign(INTERNAL, __AllAccessors__);
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/exports.ts:257]: I prefer to be explicit about it here and enumerate them all.

File: src/mono/wasm/runtime/roots.ts [hunk 2]
```diff
@@ -198,11 +198,13 @@ export class WasmRootBuffer {
...
         this._check_in_range(index);
-        return <any>Module.HEAP32[this.get_address_32(index)];
+        const offset = this.get_address_32(index);
+        return <any>Module.HEAP32[offset];
     }
 
     set(index: number, value: ManagedPointer): ManagedPointer {
-        Module.HEAP32[this.get_address_32(index)] = <any>value;
+        const offset = this.get_address_32(index);
+        Module.HEAP32[offset] = <any>value;
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/roots.ts:207]: apply the helper here too ?

**Review by pavelsavara — CHANGES_REQUESTED** [review]

File: src/mono/wasm/runtime/roots.ts (hunk 2 above)
Comment (kg) [reply src/mono/wasm/runtime/roots.ts:207]: The overhead of doing the extra calculations can add up, but I can do it if you want

Comment (pavelsavara) [reply src/mono/wasm/runtime/roots.ts:207]: Just comment on it, as you suggested in chat.

**Review by pavelsavara — APPROVED** [review]

**Pull request merged**